homeassistant/components/yolink/sensor.py

```diff
@@ -1,24 +1,106 @@
+from yolink.client import YoLinkClient
 
-from .const import DOMAIN, HOME_SUBSCRIPTION
-from .device import YoLinkDevice
-from .device_impl import YoLinkDoorSensor
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_STATE, PERCENTAGE
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import percentage
 
+from .const import (
+    ATTR_CLIENT,
+    ATTR_DEVICE,
+    ATTR_DEVICE_DOOR_SENSOR,
+    ATTR_DEVICE_TYPE,
+    ATTR_MQTT_CLIENT,
+    ATTR_PLATFORM_SENSOR,
+    DOMAIN,
+)
+from .entity import YoLinkEntity
 
-async def async_setup_entry(hass, config_entry, async_add_entities):
-    """Perform the setup for YoLink devices."""
-    devices = []
-    entities: list[YoLinkDevice] = []
-    for device in hass.data[DOMAIN][config_entry.entry_id]["devices"]:
-        if device["type"] == "DoorSensor":
-            devices.append(YoLinkDoorSensor(device, hass, config_entry))
+SENSOR_DESC_MAP: dict[str, SensorEntityDescription] = {
+    "battery": SensorEntityDescription(
+        key="battery",
+        device_class=SensorDeviceClass.BATTERY,
+        native_unit_of_measurement=PERCENTAGE,
+        name="Battery",
+        state_class=SensorStateClass.MEASUREMENT,
+    ),
+}
 
-    for device in devices:
-        entities.extend(device.entities)
+SENSOR_DEVICE_TYPE = [ATTR_DEVICE_DOOR_SENSOR]
 
-    hass.data[DOMAIN][config_entry.entry_id][HOME_SUBSCRIPTION].attachPlatformDevices(
-        Platform.SENSOR, devices
-    )
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up YoLink Sensor from a config entry."""
+    sensor_devices = [
+        device
+        for device in hass.data[DOMAIN][config_entry.entry_id][ATTR_DEVICE]
+        if device[ATTR_DEVICE_TYPE] in SENSOR_DEVICE_TYPE
+    ]
+    yl_client = hass.data[DOMAIN][config_entry.entry_id][ATTR_CLIENT]
+    yl_mqtt_client = hass.data[DOMAIN][config_entry.entry_id][ATTR_MQTT_CLIENT]
+    entities = []
+    for sensor_device in sensor_devices:
+        sensor_entities = get_sensor_entities(hass, sensor_device, yl_client)
+        if sensor_entities is not None:
+            for entity in sensor_entities:
+                entities.append(entity)
+    yl_mqtt_client.add_device_subscription(ATTR_PLATFORM_SENSOR, entities)
     async_add_entities(entities)
+
+
+def get_sensor_entities(
+    hass: HomeAssistant, device: dict, client: YoLinkClient
+) -> list[YoLinkEntity] | None:
+    """Get device entities."""
+    if device[ATTR_DEVICE_TYPE] == ATTR_DEVICE_DOOR_SENSOR:
+        return [YoLinkSensorBatteryEntity(hass, device, client)]
+    return None
+
+
+class YoLinkSensorBatteryEntity(YoLinkEntity, SensorEntity):
```

Comment:
Instead of creating an entity class for each sensor type, just pass the correct entity description for the sensor type to a single entity class when instantiating the entity.